sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12.processing.service.ts

```diff
@@ -301,12 +320,226 @@ export class IER12ProcessingService {
       };
       ier12Records.push(ier12Record);
     }
     return ier12Records;
   }
 
+  /**
+   * Get application event code (i.e current state of an application)
+   * @param applicationNumber application number.
+   * @param applicationStatus application status.
+   * @param currentDisbursementSchedule current disbursement schedule.
+   * @param studentRestrictions student restrictions.
+   * @returns application event code.
+   */
+  private async getApplicationEventCode(
+    applicationNumber: string,
+    applicationStatus: ApplicationStatus,
+    currentDisbursementSchedule: DisbursementSchedule,
+    studentRestrictions?: StudentRestriction[],
+  ): Promise<ApplicationEventCode> {
+    switch (applicationStatus) {
+      case ApplicationStatus.Assessment:
+        return this.applicationEventCodeDuringAssessment(applicationNumber);
+
+      case ApplicationStatus.Enrolment:
+        return this.applicationEventCodeDuringEnrolmentAndCompleted(
+          currentDisbursementSchedule.coeStatus,
+        );
+
+      case ApplicationStatus.Completed:
+        return this.applicationEventCodeDuringCompleted(
+          currentDisbursementSchedule,
+          studentRestrictions,
+        );
+
+      case ApplicationStatus.Cancelled:
+        return ApplicationEventCode.DISC;
+    }
+  }
+
+  /**
+   * Get application event code for an application with assessment status.
+   * @param applicationNumber application number.
+   * @returns application event code.
+   */
+  private async applicationEventCodeDuringAssessment(
+    applicationNumber: string,
+  ): Promise<ApplicationEventCode.REIA | ApplicationEventCode.ASMT> {
+    // Check if the application has more than one submissions.
+    const hasMultipleApplicationSubmissions =
+      await this.applicationService.hasMultipleApplicationSubmissions(
+        applicationNumber,
+      );
+    return hasMultipleApplicationSubmissions
+      ? ApplicationEventCode.REIA
+      : ApplicationEventCode.ASMT;
+  }
+
+  /**
+   * Get application event code for an application with enrollment/completed status.
+   * @param coeStatus coe status.
+   * @returns application event code.
+   */
+  private applicationEventCodeDuringEnrolmentAndCompleted(
+    coeStatus: COEStatus,
+  ): ApplicationEventCode.COER | ApplicationEventCode.COED {
+    // Check if the application has more than one submissions.
+    switch (coeStatus) {
+      case COEStatus.required:
+        return ApplicationEventCode.COER;
+      case COEStatus.declined:
+        return ApplicationEventCode.COED;
+    }
+  }
+
+  /**
+   * Get application event code for an application with completed status.
+   * @param currentDisbursementSchedule current disbursement schedule.
+   * @param studentRestrictions student restrictions.
+   * @returns application event code.
+   */
+  private async applicationEventCodeDuringCompleted(
+    currentDisbursementSchedule: DisbursementSchedule,
+    studentRestrictions?: StudentRestriction[],
```

Comment:
The methods can receive the object as specific as they are needed.
For instance, `studentRestrictions` is provided as a parameter when probably only the actual restrictions array is needed, and from that only `actionType` is actually used. Please let me know if I am missing something.
Tha same is probably the case for the `DisbursementSchedule`.
IMO having more specific parameters will help building the unit tests also.